Paco-Server/src/com/google/sampling/experiential/server/CloudSQLDaoImpl.java

```diff
@@ -0,0 +1,418 @@
+        ((ExpressionList) outputInsert.getItemsList()).getExpressions()
+                                                      .add(whatAnswer != null ? new StringValue(quote(whatAnswer))
+                                                                              : null);
+      }
+
+      statementCreateEvent = conn.prepareStatement(eventInsert.toString());
+      statementCreateEvent.execute();
+
+      statementCreateEventOutput = conn.prepareStatement(outputInsert.toString());
+      statementCreateEventOutput.executeBatch();
+
+      conn.commit();
+
+      retVal = true;
+    } catch (SQLException e) {
+      log.info("While inserting event : " + e);
+    } catch (Exception e) {
+      log.info("While inserting event -Gen : " + e);
+    } finally {
+      try {
+        if (statementCreateEventOutput != null) {
+          statementCreateEventOutput.close();
+        }
+        if (statementCreateEvent != null) {
+          statementCreateEvent.close();
+        }
+        if (conn != null) {
+          conn.close();
+        }
+      } catch (SQLException ex1) {
+        log.warning("While closing conn for insert event" + ex1);
+      }
+    }
+    return retVal;
+  }
+
+  @Override
+  public List<EventDAO> getEvents(String query) throws SQLException {
+    List<EventDAO> evtList = Lists.newArrayList();
+    EventDAO event = null;
+    Connection conn = null;
+    Map<Long, EventDAO> eventMap = null;
+    PreparedStatement statementSelectEvent = null;
+    ResultSet rs = null;
+
+    try {
+      conn = CloudSQLConnectionManager.getInstance().getConnection();
+      statementSelectEvent = conn.prepareStatement(query, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+      statementSelectEvent.setFetchSize(Integer.MIN_VALUE);
+      Long stTime = System.nanoTime();
+      rs = statementSelectEvent.executeQuery();
+
+      log.info("step 1 query " + query + "took" + (System.nanoTime() - stTime));
+
+      // to maintain the insertion order
+      eventMap = Maps.newLinkedHashMap();
+      Long dt = System.nanoTime();
+      if (rs != null) {
+        evtList = Lists.newArrayList();
+        while (rs.next()) {
+          event = createEvent(rs);
+          EventDAO oldEvent = eventMap.get(event.getId());
+          if (oldEvent == null) {
+            eventMap.put(event.getId(), event);
+          } else {
+            // Will go through following when the query contains text in
+            // ('a','b','v') or answer in(....)
+            List<WhatDAO> oldOut = oldEvent.getWhat();
+            List<WhatDAO> newOut = event.getWhat();
+            // add all new output to the existing event in map
+            oldOut.addAll(newOut);
+          }
+        }
+      }
+      // TODO step 2 sometimes takes 4 times longer to execute than the query.
+      // Not sure why??
+      log.info("step 2 proc took" + (System.nanoTime() - dt));
```

Comment:
Please add " nanoseconds" to the log event so we remember it as nanos. Or, I learned a nifty new trick from jdk 7, TimeUtil.NanoSeconds.toMillis(System.nanoTime() - dt) to put it into milliseconds (easier to understand).